### PR TITLE
Improve `Persistent_data`

### PR DESCRIPTION
**Pull request by smondet**
The main points:

- Cache “pre-fetching” (dumb but useful)
- Cache invalidation together with target archival (i.e. not only in the cache but also in the DB)
- Killing of orphan targets

There is also better logging of some functions and whole HTTP requests.

---
This change is [Reviewable](https://reviewable.io/reviews/hammerlab/ketrew/419)

### CODE REVIEW

**Comment by smondet** [comment]
I forgot to do one thing, not ready for merge :)

This makes `find_all_orphans` loop forever:

```ocaml
let () =
  let open Printf in
  begin
    let configuration =
      Ketrew.Configuration.(
        client "https://127.0.0.1:8443" ~token:"nekot" |> create
      ) in
    let t1 = Ketrew_pure.Target.create ~id:"t1" ~depends_on:["t2"] () in
    let t2 = Ketrew_pure.Target.create ~id:"t2" ~depends_on:["t1"] () in
    Ketrew.Client.as_client ~configuration ~f:begin fun ~client ->
      Ketrew.Client.add_targets client [t1; t2]
    end
  end
  |> Lwt_main.run
  |> begin function
  | `Ok () -> ()
  | `Error e -> eprintf "Error: %s\n%!" (Ketrew.Error.to_string e); failwith "ERROR"
  end

```

File: src/lib/client.mli [hunk 1]
```diff
@@ -154,6 +154,19 @@ val restart: t ->
...
+  Ketrew_pure.Target.t list ->
+  (unit,
+   [> `Client of Error.t
+   | `Database of Trakeva.Error.t
+   | `Database_unavailable of string
+   | `Missing_data of string
+   | `Target of [> `Deserilization of string ] ])
+    Deferred_result.t
+(** Submit a list of targets to the server/engine. This is the low-level API
+    deadling with {!Ketrew_pure.Target.t} values; for use in the EDSL please
```
Comment (ihodes) [inline src/lib/client.mli]: -> dealing

File: src/lib/persistent_data.ml [hunk 2]
```diff
@@ -746,27 +757,136 @@ let get_target t id =
...
+    check_target t
+  end
+
+(** Go through the [to_gc] list and remove those from the list-of-i cache, the
+    from the cache-table, and use {!With_database.get_stored_target} to make
+    sure {!With_database.clean_up_finished} is also called.
+
+    In the function {!all_visible_targets}, we call
+    {!perform_garbage_collection} inside a {!Lwt.async} call, so that garbage
+    collection does not bloc the UIs that depend on it.
```
Comment (ihodes) [inline src/lib/persistent_data.ml]: -> block

File: src/lib/persistent_data.ml [hunk 3]
```diff
@@ -746,27 +757,136 @@ let get_target t id =
...
+   ]}
+*)
+let rec should_garbage_collect ?(acc = []) t stored_target =
+  let check_target trgt =
+    let threshold = With_database.get_archival_threshold t.db in
+    let now = Time.now () in
+    begin match Target.(state trgt |> State.finished_time) with
+    | None -> return `No
+    | Some time when time +. threshold > now ->
+      (* fnished, but not old enough to be garbage collected *)
```
Comment (ihodes) [inline src/lib/persistent_data.ml]: -> finished

File: src/lib/persistent_data.ml [hunk 4]
```diff
@@ -746,27 +757,136 @@ let get_target t id =
...
+  | `Pointer id ->
+    get_stored_target t id
+    >>= fun points_to ->
+    let item = id, `Points_to (Target.Stored_target.id points_to) in
+    should_garbage_collect ~acc:(item  :: acc) t points_to
+  | `Target t ->
+    check_target t
+  end
+
+(** Go through the [to_gc] list and remove those from the list-of-i cache, the
```
Comment (ihodes) [inline src/lib/persistent_data.ml]: don't need the final "the"

File: src/lib/persistent_data.ml [hunk 5]
```diff
@@ -746,27 +757,136 @@ let get_target t id =
...
+            return (`P p, `Min (min t mi), `Max (max t ma), `Errors errors)
+          end
+        | `Error e ->
+          return (`P p, `Min mi, `Max ma, `Errors (e :: errors))
+      end
+    end
+  >>= fun (`P pointers, `Min min_time, `Max max_time, `Errors errors) ->
+  log_markup t Display_markup.([
+      "function", text "perform_garbage_collection";
+      "to-garabage-collect", (
```
Comment (ihodes) [inline src/lib/persistent_data.ml]: garabage -> garbage

File: src/lib/persistent_data.ml [hunk 6]
```diff
@@ -746,27 +757,136 @@ let get_target t id =
...
+            | less ->
+              description_list (List.map to_gc ~f:(fun (id, why) ->
+                  id, (match why with
+                    | `Points_to id -> textf "-> %s" id
+                    | `Time t -> concat [textf "finsihed on "; date t]
+                    )))
+          );
+        ]
+      );
+      "start_date", date start_date;
```
Comment (ihodes) [inline src/lib/persistent_data.ml]: -> start-date

**Comment by ihodes** [comment]
LGTM! Few comments on some typos, but looking forward to having this in Ketrew.

**Pull request merged**